cannon/cmd/run.go

```diff
@@ -166,23 +179,37 @@ func (p *ProcessPreimageOracle) Start() error {
 	if p.cmd == nil {
 		return nil
 	}
-	return p.cmd.Start()
+	err := p.cmd.Start()
+	go p.wait()
+	return err
 }
 
 func (p *ProcessPreimageOracle) Close() error {
 	if p.cmd == nil {
 		return nil
 	}
+	// We first check if the process has already exited before signaling.
+	// Note: This is a teeny bit racy since the process could exit after the check
+	// above and another process is assigned the same pid.
+	select {
+	case err := <-p.waitErr:
+		return err
+	case <-time.After(time.Second * 1):
+		// give the wait goroutine time to reap process
+	}
 	_ = p.cmd.Process.Signal(os.Interrupt)
-	// Go 1.20 feature, to introduce later
-	//p.cmd.WaitDelay = time.Second * 10
+	return <-p.waitErr
```

Comment:
Why do we need to wait before interrupting here? Wouldn't an interrupt to an exited process just be a no-op (especially if we ignore the error)?

And I would have expected that go would hold a reference to the actual child process it spawned rather than having to repeatedly look it up by pid which should avoid the race condition. Is it not doing that for some reason?

Reply:
you're right, interrupting an exited process is a no-op. I misread the stdlib as doing otherwise.